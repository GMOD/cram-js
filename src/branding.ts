export type Brand<
  Base,
  Branding,
  ReservedName extends string = '__type__',
> = Base & { [K in ReservedName]: Branding } & { __witness__: Base }

function identity<TIn, TOut>() {
  return (item: TIn) => item as any as TOut
}

export type UInt8 = Brand<number, 'uint8'>
export const makeUInt8 = identity<number, UInt8>()

export function isUint8(n: number): n is UInt8 {
  if (!Number.isSafeInteger(n)) {
    return false
  }
  return n >= 0 && n <= 255
}

export function ensureUint8(n: number): UInt8 {
  if (!isUint8(n)) {
    throw new Error(n + ' is not uint8')
  }
  return n
}

export type Int8 = Brand<number, 'int8'>
export const assertInt8 = identity<number, Int8>()

export function isInt8(n: number): n is Int8 {
  if (!Number.isSafeInteger(n)) {
    return false
  }
  return n >= -128 && n <= 127
}

export function ensureInt8(n: number): Int8 {
  if (!isInt8(n)) {
    throw new Error(n + ' is not int8')
  }
  return n
}

export type Int32 = Brand<number, 'int32'>

export function assertInt32(n: number): Int32 {
  return n as Int32
}

const minInt32 = -2147483648
const maxInt32 = 2147483647

export function isInt32(n: number): n is Int32 {
  return Number.isSafeInteger(n) && n >= minInt32 && n <= maxInt32
}

export function ensureInt32(n: number): Int32 {
  if (!isInt32(n)) {
    throw new Error(n + ' is not int32')
  }
  return n
}

export type Int64 = Brand<bigint, 'int64'>

export function assertInt64(n: bigint): Int64 {
  return n as Int64
}

export function isInt64n(n: number): boolean {
  return Number.isSafeInteger(n)
}

const minInt64 = BigInt('-922372036854775808')
const maxInt64 = BigInt(' 9223372036854775807')

export function isInt64(n: bigint): n is Int64 {
  return n >= minInt64 && n <= maxInt64
}

export function ensureInt64(n: bigint): Int64 {
  if (!isInt64(n)) {
    throw new Error(n + ' is not int64')
  }
  return n
}

export function ensureInt64n(n: number): Int64 {
  if (!isInt64n(n)) {
    throw new Error(n + ' is not int64')
  }
  return assertInt64(BigInt(n))
}

export function addInt32(a: Int32, b: Int32) {
  return ensureInt32(a + b)
}

export function incrementInt32(a: Int32) {
  return addInt32(a, assertInt32(1))
}

export function decrementInt32(a: Int32) {
  return addInt32(a, assertInt32(-1))
}

export function subtractInt32(a: Int32, b: Int32) {
  return ensureInt32(a - b)
}
