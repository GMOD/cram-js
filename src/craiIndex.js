const AbortablePromiseCache = require('abortable-promise-cache').default
const QuickLRU = require('quick-lru')

const { promisify } = require('es6-promisify')
const zlib = require('zlib')

const gunzip = promisify(zlib.gunzip)

const { open } = require('./io')
const { CramMalformedError } = require('./errors')

const BAI_MAGIC = 21578050 // BAI\1

class Slice {
  constructor(args) {
    Object.assign(this, args)
  }

  toString() {
    return `${this.start}:${this.span}:${this.containerStart}:${this.sliceStart}:${this.sliceBytes}`
  }
}

function addRecordToIndex(index, record) {
  if (record.some(el => el === undefined)) {
    throw new CramMalformedError('invalid .crai index file')
  }

  const [seqId, start, span, containerStart, sliceStart, sliceBytes] = record

  if (!index[seqId]) index[seqId] = []

  index[seqId].push(
    new Slice({
      start,
      span,
      containerStart,
      sliceStart,
      sliceBytes,
    }),
  )
}
class CraiIndex {
  // A CRAM index (.crai) is a gzipped tab delimited file containing the following columns:
  // 1. Sequence id
  // 2. Alignment start
  // 3. Alignment span
  // 4. Container start byte position in the file
  // 5. Slice start byte position in the container data (‘blocks’)
  // 6. Slice size in bytes
  // Each line represents a slice in the CRAM file. Please note that all slices must be listed in index file.

  /**
   *
   * @param {object} args
   * @param {string} [args.path]
   * @param {string} [args.url]
   * @param {FileHandle} [args.filehandle]
   */
  constructor(args) {
    const filehandle = open(args.url, args.path, args.filehandle)
    this._parseCache = new AbortablePromiseCache({
      cache: new QuickLRU({ maxSize: 1 }),
      fill: (data, signal) => this.parseIndex({ signal }),
    })
    this.readFile = filehandle.readFile.bind(filehandle)
  }

  parseIndex() {
    const index = {}
    return this.readFile()
      .then(data => {
        if (data[0] === 31 && data[1] === 139) return gunzip(data)
        return data
      })
      .then(uncompressedBuffer => {
        if (
          uncompressedBuffer.length > 4 &&
          uncompressedBuffer.readUInt32LE(0) === BAI_MAGIC
        ) {
          throw new CramMalformedError(
            'invalid .crai index file. note: file appears to be a .bai index. this is technically legal but please open a github issue if you need support',
          )
        }
        // interpret the text as regular ascii, since it is
        // supposed to be only digits and whitespace characters
        // this is written in a deliberately low-level fashion for performance,
        // because some .crai files can be pretty large.
        let currentRecord = []
        let currentString = ''
        for (let i = 0; i < uncompressedBuffer.length; i += 1) {
          const charCode = uncompressedBuffer[i]
          if (
            (charCode >= 48 && charCode <= 57) /* 0-9 */ ||
            (!currentString && charCode === 45) /* leading - */
          ) {
            currentString += String.fromCharCode(charCode)
          } else if (charCode === 9 /* \t */) {
            currentRecord.push(Number.parseInt(currentString, 10))
            currentString = ''
          } else if (charCode === 10 /* \n */) {
            currentRecord.push(Number.parseInt(currentString, 10))
            currentString = ''
            addRecordToIndex(index, currentRecord)
            currentRecord = []
          } else if (charCode !== 13 /* \r */ && charCode !== 32 /* space */) {
            // if there are other characters in the file besides
            // space and \r, something is wrong.
            throw new CramMalformedError('invalid .crai index file')
          }
        }

        // if the file ends without a \n, we need to flush our buffers
        if (currentString) {
          currentRecord.push(Number.parseInt(currentString, 10))
        }
        if (currentRecord.length === 6) {
          addRecordToIndex(index, currentRecord)
        }

        // sort each of them by start
        Object.entries(index).forEach(([seqId, ent]) => {
          index[seqId] = ent.sort(
            (a, b) => a.start - b.start || a.span - b.span,
          )
        })
        return index
      })
  }

  getIndex(opts = {}) {
    return this._parseCache.get('index', null, opts.signal)
  }

  /**
   * @param {number} seqId
   * @returns {Promise} true if the index contains entries for
   * the given reference sequence ID, false otherwise
   */
  async hasDataForReferenceSequence(seqId) {
    return !!(await this.getIndex())[seqId]
  }

  /**
   * fetch index entries for the given range
   *
   * @param {number} seqId
   * @param {number} queryStart
   * @param {number} queryEnd
   *
   * @returns {Promise} promise for
   * an array of objects of the form
   * `{start, span, containerStart, sliceStart, sliceBytes }`
   */
  async getEntriesForRange(seqId, queryStart, queryEnd) {
    const seqEntries = (await this.getIndex())[seqId]
    if (!seqEntries) {
      return []
    }

    const compare = entry => {
      const entryStart = entry.start
      const entryEnd = entry.start + entry.span
      if (entryStart >= queryEnd) return -1 // entry is ahead of query
      if (entryEnd <= queryStart) return 1 // entry is behind query
      return 0 // entry overlaps query
    }
    const bins = []
    for (let i = 0; i < seqEntries.length; i += 1) {
      if (compare(seqEntries[i]) === 0) {
        bins.push(seqEntries[i])
      }
    }
    return bins
  }
}

module.exports = CraiIndex
