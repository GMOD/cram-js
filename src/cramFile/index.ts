import CramFile from './file'

export default CramFile
