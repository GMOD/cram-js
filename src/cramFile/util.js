const md5 = require('md5')
const { CramBufferOverrunError } = require('../errors')

module.exports = {
  itf8Size(v) {
    if (!(v & ~0x7f)) return 1
    if (!(v & ~0x3fff)) return 2
    if (!(v & ~0x1fffff)) return 3
    if (!(v & ~0xfffffff)) return 4
    return 5
  },

  parseItf8(buffer, initialOffset) {
    let offset = initialOffset
    const countFlags = buffer[offset]
    let result
    if (countFlags < 0x80) {
      result = countFlags
      offset += 1
    } else if (countFlags < 0xc0) {
      result = ((countFlags << 8) | buffer[offset + 1]) & 0x3fff
      offset += 2
    } else if (countFlags < 0xe0) {
      result =
        ((countFlags << 16) | (buffer[offset + 1] << 8) | buffer[offset + 2]) &
        0x1fffff
      offset += 3
    } else if (countFlags < 0xf0) {
      result =
        ((countFlags << 24) |
          (buffer[offset + 1] << 16) |
          (buffer[offset + 2] << 8) |
          buffer[offset + 3]) &
        0x0fffffff
      offset += 4
    } else {
      result =
        ((countFlags & 0x0f) << 28) |
        (buffer[offset + 1] << 20) |
        (buffer[offset + 2] << 12) |
        (buffer[offset + 3] << 4) |
        (buffer[offset + 4] & 0x0f)
      // x=((0xff & 0x0f)<<28) | (0xff<<20) | (0xff<<12) | (0xff<<4) | (0x0f & 0x0f);
      // TODO *val_p = uv < 0x80000000UL ? uv : -((int32_t) (0xffffffffUL - uv)) - 1;
      offset += 5
    }
    if (offset > buffer.length) {
      throw new CramBufferOverrunError(
        'Attempted to read beyond end of buffer; this file seems truncated.',
      )
    }
    return [result, offset - initialOffset]
  },

  // parseLtf8(buffer, initialOffset) {
  //   let offset = initialOffset
  //   const countFlags = buffer[offset]
  //   let result
  //   if (countFlags < 0x80) {
  //     result = countFlags
  //     offset += 1
  //   } else if (countFlags < 0xc0) {
  //     result = ((buffer[offset] << 8) | buffer[offset + 1]) & 0x3fff
  //     offset += 2
  //   } else if (countFlags < 0xe0) {
  //     result =
  //       ((buffer[offset] << 16) |
  //         (buffer[offset + 1] << 8) |
  //         buffer[offset + 2]) &
  //       0x1fffff
  //     offset += 3
  //   } else if (countFlags < 0xf0) {
  //     result =
  //       ((buffer[offset] << 24) |
  //         (buffer[offset + 1] << 16) |
  //         (buffer[offset + 2] << 8) |
  //         buffer[offset + 3]) &
  //       0x0fffffff
  //     offset += 4
  //   } else if (countFlags < 0xf8) {
  //     result =
  //       ((buffer[offset] & 15) * Math.pow(2,32) + (buffer[offset + 1] << 24)) |
  //       ((buffer[offset + 2] << 16) |
  //         (buffer[offset + 3] << 8) |
  //         buffer[offset + 4])
  //     // TODO *val_p = uv < 0x80000000UL ? uv : -((int32_t) (0xffffffffUL - uv)) - 1;
  //     offset += 5
  //   } else if (countFlags < 0xfc) {
  //     result =
  //       ((((buffer[offset] & 7) << 8) | buffer[offset + 1]) * Math.pow(2,32) +
  //         (buffer[offset + 2] << 24)) |
  //       ((buffer[offset + 3] << 16) |
  //         (buffer[offset + 4] << 8) |
  //         buffer[offset + 5])
  //     offset += 6
  //   } else if (countFlags < 0xfe) {
  //     result =
  //       ((((buffer[offset] & 3) << 16) |
  //         (buffer[offset + 1] << 8) |
  //         buffer[offset + 2]) *
  //         Math.pow(2,32) +
  //         (buffer[offset + 3] << 24)) |
  //       ((buffer[offset + 4] << 16) |
  //         (buffer[offset + 5] << 8) |
  //         buffer[offset + 6])
  //     offset += 7
  //   } else if (countFlags < 0xff) {
  //     result = Long.fromBytesBE(buffer.slice(offset + 1, offset + 8))
  //     if (
  //       result.greaterThan(Number.MAX_SAFE_INTEGER) ||
  //       result.lessThan(Number.MIN_SAFE_INTEGER)
  //     )
  //       throw new CramUnimplementedError('integer overflow')
  //     result = result.toNumber()
  //     offset += 8
  //   } else {
  //     result = Long.fromBytesBE(buffer.slice(offset + 1, offset + 9))
  //     if (
  //       result.greaterThan(Number.MAX_SAFE_INTEGER) ||
  //       result.lessThan(Number.MIN_SAFE_INTEGER)
  //     )
  //       throw new CramUnimplementedError('integer overflow')
  //     result = result.toNumber()
  //     offset += 9
  //   }
  //   return [result, offset - initialOffset]
  // },

  parseItem(buffer, parser, startBufferPosition = 0, startFilePosition = 0) {
    const { offset, result } = parser.parse(buffer)
    result._endPosition = offset + startFilePosition
    result._size = offset - startBufferPosition
    return result
  },

  // this would be nice as a decorator, but i'm a little worried about
  // babel support for it going away or changing.
  // memoizes a method in the stupidest possible way, with no regard for the
  // arguments.  actually, this only works on methods that take no arguments
  tinyMemoize(_class, methodName) {
    const method = _class.prototype[methodName]
    const memoAttrName = `_memo_${methodName}`
    _class.prototype[methodName] = function _tinyMemoized() {
      if (!(memoAttrName in this)) {
        const res = method.call(this)
        this[memoAttrName] = res
        Promise.resolve(res).catch(err => {
          delete this[memoAttrName]
          throw err
        })
      }
      return this[memoAttrName]
    }
  },

  sequenceMD5(seq) {
    return md5(seq.toUpperCase().replace(/[^\x21-\x7e]/g, ''))
  },
}
