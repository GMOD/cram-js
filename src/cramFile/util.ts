import md5 from 'md5'

import { CramBufferOverrunError } from './codecs/getBits'
import { longFromBytesToUnsigned } from './long'

export function itf8Size(v: number) {
  if (!(v & ~0x7f)) {
    return 1
  }
  if (!(v & ~0x3fff)) {
    return 2
  }
  if (!(v & ~0x1fffff)) {
    return 3
  }
  if (!(v & ~0xfffffff)) {
    return 4
  }
  return 5
}

export function parseItf8(buffer: Uint8Array, initialOffset: number) {
  let offset = initialOffset
  const countFlags = buffer[offset]!
  let result: number
  if (countFlags < 0x80) {
    result = countFlags
    offset = offset + 1
  } else if (countFlags < 0xc0) {
    result = ((countFlags << 8) | buffer[offset + 1]!) & 0x3fff
    offset = offset + 2
  } else if (countFlags < 0xe0) {
    result =
      ((countFlags << 16) | (buffer[offset + 1]! << 8) | buffer[offset + 2]!) &
      0x1fffff
    offset = offset + 3
  } else if (countFlags < 0xf0) {
    result =
      ((countFlags << 24) |
        (buffer[offset + 1]! << 16) |
        (buffer[offset + 2]! << 8) |
        buffer[offset + 3]!) &
      0x0fffffff
    offset = offset + 4
  } else {
    result =
      ((countFlags & 0x0f) << 28) |
      (buffer[offset + 1]! << 20) |
      (buffer[offset + 2]! << 12) |
      (buffer[offset + 3]! << 4) |
      (buffer[offset + 4]! & 0x0f)
    // x=((0xff & 0x0f)<<28) | (0xff<<20) | (0xff<<12) | (0xff<<4) | (0x0f & 0x0f);
    // TODO *val_p = uv < 0x80000000UL ? uv : -((int32_t) (0xffffffffUL - uv)) - 1;
    offset = offset + 5
  }
  if (offset > buffer.length) {
    throw new CramBufferOverrunError(
      'Attempted to read beyond end of buffer; this file seems truncated.',
    )
  }
  return [result, offset - initialOffset] as const
}

export function parseLtf8(buffer: Uint8Array, initialOffset: number) {
  const dataView = new DataView(buffer.buffer)
  let offset = initialOffset
  const countFlags = buffer[offset]!
  let n: number
  if (countFlags < 0x80) {
    n = countFlags
    offset += 1
  } else if (countFlags < 0xc0) {
    n = ((buffer[offset]! << 8) | buffer[offset + 1]!) & 0x3fff
    offset += 2
  } else if (countFlags < 0xe0) {
    n =
      ((buffer[offset]! << 16) |
        (buffer[offset + 1]! << 8) |
        buffer[offset + 2]!) &
      0x1fffff
    n = ((countFlags & 63) << 16) | dataView.getUint16(offset + 1, true)
    offset += 3
  } else if (countFlags < 0xf0) {
    n =
      ((buffer[offset]! << 24) |
        (buffer[offset + 1]! << 16) |
        (buffer[offset + 2]! << 8) |
        buffer[offset + 3]!) &
      0x0fffffff
    offset += 4
  } else if (countFlags < 0xf8) {
    n =
      ((buffer[offset]! & 15) * 2 ** 32 + (buffer[offset + 1]! << 24)) |
      ((buffer[offset + 2]! << 16) |
        (buffer[offset + 3]! << 8) |
        buffer[offset + 4]!)
    // TODO *val_p = uv < 0x80000000UL ? uv : -((int32_t) (0xffffffffUL - uv)) - 1;
    offset += 5
  } else if (countFlags < 0xfc) {
    n =
      ((((buffer[offset]! & 7) << 8) | buffer[offset + 1]!) * 2 ** 32 +
        (buffer[offset + 2]! << 24)) |
      ((buffer[offset + 3]! << 16) |
        (buffer[offset + 4]! << 8) |
        buffer[offset + 5]!)
    offset += 6
  } else if (countFlags < 0xfe) {
    n =
      ((((buffer[offset]! & 3) << 16) |
        (buffer[offset + 1]! << 8) |
        buffer[offset + 2]!) *
        2 ** 32 +
        (buffer[offset + 3]! << 24)) |
      ((buffer[offset + 4]! << 16) |
        (buffer[offset + 5]! << 8) |
        buffer[offset + 6]!)
    offset += 7
  } else if (countFlags < 0xff) {
    n = longFromBytesToUnsigned(buffer.subarray(offset + 1, offset + 8))
    offset += 8
  } else {
    n = longFromBytesToUnsigned(buffer.subarray(offset + 1, offset + 9))
    offset += 9
  }
  return [n, offset - initialOffset] as const
}

export function parseItem<T>(
  buffer: Uint8Array,
  parser: (buffer: Uint8Array, offset: number) => { offset: number; value: T },
  startBufferPosition = 0,
  startFilePosition = 0,
) {
  const { offset, value } = parser(buffer, startBufferPosition)
  return {
    ...value,
    _endPosition: offset + startFilePosition,
    _size: offset - startBufferPosition,
  }
}

// this would be nice as a decorator, but i'm a little worried about babel
// support for it going away or changing. memoizes a method in the stupidest
// possible way, with no regard for the arguments.  actually, this only works
// on methods that take no arguments
export function tinyMemoize(_class: any, methodName: any) {
  const method = _class.prototype[methodName]
  const memoAttrName = `_memo_${methodName}`
  _class.prototype[methodName] = function _tinyMemoized() {
    if (!(memoAttrName in this)) {
      const res = method.call(this)
      this[memoAttrName] = res
      Promise.resolve(res).catch(() => {
        delete this[memoAttrName]
      })
    }
    return this[memoAttrName]
  }
}

export function sequenceMD5(seq: string) {
  return md5(seq.toUpperCase().replaceAll(/[^\u0021-\u007e]/g, ''))
}
