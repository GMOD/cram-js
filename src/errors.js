export class CramError extends Error {}

/** Error caused by encountering a part of the CRAM spec that has not yet been implemented */
export class CramUnimplementedError extends Error {}

/** An error caused by malformed data.  */
export class CramMalformedError extends CramError {}

/**
 * An error caused by attempting to read beyond the end of the defined data.
 */
export class CramBufferOverrunError extends CramMalformedError {}

/**
 * An error caused by data being too big, exceeding a size limit.
 */
export class CramSizeLimitError extends CramError {}

/**
 * An invalid argument was supplied to a cram-js method or object.
 */
export class CramArgumentError extends CramError {}
