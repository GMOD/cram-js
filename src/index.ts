import CramFile from './cramFile'
import IndexedCramFile from './indexedCramFile'
import CraiIndex from './craiIndex'

export { CramFile, IndexedCramFile, CraiIndex }
