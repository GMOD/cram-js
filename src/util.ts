function sum(array: Uint8Array[]) {
  let sum = 0
  for (const entry of array) {
    sum += entry.length
  }
  return sum
}
export function concatUint8Array(args: Uint8Array[]) {
  const mergedArray = new Uint8Array(sum(args))
  let offset = 0
  for (const entry of args) {
    mergedArray.set(entry, offset)
    offset += entry.length
  }
  return mergedArray
}
