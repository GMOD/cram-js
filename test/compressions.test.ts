//@ts-nocheck
import { test, expect } from 'vitest'
import { testDataFile } from './lib/util'
import { dumpWholeFile } from './lib/dumpFile'
import { CramFile } from '../src/index'
import { FetchableSmallFasta } from './lib/fasta'

test('lzma', async () => {
  const fasta = new FetchableSmallFasta(testDataFile('ce.fa'))
  const seqFetch = fasta.fetch.bind(fasta)
  const file = new CramFile({
    filehandle: testDataFile('hts-specs/0903_comp_lzma.cram'),
    seqFetch,
  })
  const fileData = await dumpWholeFile(file)
  const feat = fileData[2].data[1].features[0]
  const hardClip = feat.readFeatures[0]
  expect(hardClip).toMatchSnapshot()
})

test('bzip2', async () => {
  const fasta = new FetchableSmallFasta(testDataFile('ce.fa'))
  const seqFetch = fasta.fetch.bind(fasta)
  const file = new CramFile({
    filehandle: testDataFile('hts-specs/0902_comp_bz2.cram'),
    seqFetch,
  })
  const fileData = await dumpWholeFile(file)
  const feat = fileData[2].data[1].features[0]
  const hardClip = feat.readFeatures[0]
  expect(hardClip).toMatchSnapshot()
})
