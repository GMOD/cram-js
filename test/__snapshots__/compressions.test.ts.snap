// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bzip2 1`] = `
{
  "code": "X",
  "data": 1,
  "pos": 51,
  "ref": "A",
  "refPos": 1050,
  "sub": "G",
}
`;

exports[`lzma 1`] = `
{
  "code": "X",
  "data": 1,
  "pos": 51,
  "ref": "A",
  "refPos": 1050,
  "sub": "G",
}
`;
